gobblin-core-base/src/main/java/gobblin/writer/AsyncWriterManager.java

```diff
@@ -0,0 +1,519 @@
+    if (isInstrumentationEnabled()) {
+      this.dataWriterTimer = Optional.of(this.metricContext.timer(MetricNames.DataWriterMetrics.WRITE_TIMER));
+    } else {
+      this.dataWriterTimer = Optional.absent();
+    }
+  }
+
+
+  protected AsyncWriterManager(Config config,
+      long commitTimeoutInNanos,
+      long commitStepWaitTimeMillis,
+      double failureAllowance,
+      boolean retriesEnabled,
+      int numRetries,
+      int maxOutstandingWrites,
+      @NonNull AsyncDataWriter asyncDataWriter)
+  {
+    this.closer = Closer.create();
+    State state = ConfigUtils.configToState(config);
+    this.instrumentationEnabled = GobblinMetrics.isEnabled(state);
+    this.metricContext = this.closer.register(Instrumented.getMetricContext(state, asyncDataWriter.getClass()));
+
+    regenerateMetrics();
+
+    this.commitTimeoutInNanos = commitTimeoutInNanos;
+    this.commitStepWaitTimeMillis = commitStepWaitTimeMillis;
+    Preconditions.checkArgument((failureAllowance <= 1.0 && failureAllowance >= 0), "Failure Allowance must be a ratio between 0 and 1");
+    this.failureAllowance = failureAllowance;
+    this.retriesEnabled = retriesEnabled;
+    if (this.retriesEnabled)
+    {
+      retryQueue = Optional.of(new LinkedBlockingQueue<Attempt>());
+      this.numRetries = numRetries;
+      Thread retryThread = new Thread(new RetryRunner());
+      retryThread.setDaemon(true);
+      retryThread.setName("AsyncWriterRetryThread");
+      retryThread.start();
+    }
+    else
+    {
+      retryQueue = Optional.absent();
+      this.numRetries = 0;
+    }
+    this.maxOutstandingWrites = maxOutstandingWrites;
+    this.asyncDataWriter = asyncDataWriter;
+    this.closer.register(asyncDataWriter);
+
+
+    this.writeCallback = new WriteCallback() {
+      @Override
+      public void onSuccess(WriteResponse writeResponse) {
+        recordsSuccess.mark();
+        log.debug("Received success", writeResponse.getStringResponse());
+      }
+
+      @Override
+      public void onFailure(Throwable throwable) {
+        recordsFailed.mark();
+      }
+    };
+  }
+
+  @Override
+  public void write(final D record)
+      throws IOException {
+    /**
+     * A few ways to achieve max concurrency.
+     * a) Keep a counter that increments and decrements.
+     * Problem: you have to spin lock on the counter on a write... may not be too bad.
+     * Dangers: If you don't decrement the counter on corner cases, you may have a false sense of how many outstanding
+     * requests you have.
+     * b) Use a bounded queue to store call attempts, write the returned futures into it, remove the future on callback success.
+     * Problem: More state to keep in memory.
+     * How do you deal with retries.. how do you ensure there is no starvation on retries?
+     *
+     * Doing a) for now.
+     */
+
+    maybeThrow();
+    int spinNum = 0;
+    while (recordsIn.getCount() - (recordsSuccess.getCount() + recordsFailed.getCount()) > maxOutstandingWrites)
```

Comment:
i'd refactor the check in a method. It is conceivable that we may want to add a size-based limit in the future (that may be more useful than a count-based limit).